### ISSUE TITLE
libc++ and __builtin_assume behavior don't mix well

### ISSUE THREAD

**Issue by halayli** [llvm/llvm-project#59775]
depending on macro definitions like _LIBCPP_ENABLE_ASSERTIONS, libc++ can end up using `__builtin_assume` and if the assumption is false at compiler time it results in 0 instructions being generated from that point forward without any error or warning.

The example below illustrates the issue. This is a case I hit but in a much less obvious way since input's size was coming from a series of template expansions.

This generates zero instructions, not even the normal `xor eax, eax;ret` because 22 != 256 which ends up with __builtin_assume(false) and anything below that is unreachable. This can result in half compiled functions and again without a single compiler warning or error.

Example:

https://godbolt.org/z/Kq5qq4Tn8 

```

#include <array>
#include <span>

int main()
{
    std::array<int, 256> chunk{};
    span<int, 256> input{chunk.data(), 22};

    return 0;
}
```

This can be categorized as diagnostics bug by llvm but I think libc++ team should know about it given `__builtin_assume`'s behavior.

I hit this issue after using libc++ from llvm-15's release but __builtin_assume has always behaved this way.

**Comment by mordante** [comment]
This seems to be a bug in your code. When you specify both a non-dynamic extend _and_ a count, both need to be equal. Not doing so violates the precondition http://eel.is/c++draft/views.contiguous#span.cons-4.3, which means your code contains UB.

Compiling with `-D_LIBCPP_ENABLE_ASSERTIONS` will result in code generated and an unconditional call to `__libcpp_verbose_abort`.

I think it would be nice to have a way for clang-tidy to warn about this construct. I might want to look at that.

**Comment by philnik777** [comment]
This is the same behaviour as core-wording UB, so I don't see a problem with it. Adding `-D_LIBCPP_ENABLE_ASSERTIONS` even finds this UB, so it's quite easy to track down if one of the assertions is violated.

**Comment by halayli** [comment]
Yes It's definitely a bug in the code. I am reporting this behavior from a usability / user experience perspective. A clang-tidy warning would have been great or a clang warning about `__builtin_assume(false)`.

**Comment by halayli** [comment]
Just to clarify, libc++ is not at fault here. It's a UB after all. It's just that the consequences of `__builtin_assume(false)` are so obscure, and silent. The ideal fix is probably adding a new warn diag for `__builtin_assume(false)`. I wouldn't label it as a libc++ issue, but I thought it might be useful for libc++ team to know about it.

**Comment by philnik777** [comment]
Do you mean something like "`__builtin_assume` condition is always false. If that is intended, use `__builtin_unreachable` instead."?

**Comment by halayli** [comment]
Exactly, and If it's not intentional the user has enough information to go from here.
